src/Authoring/WinRT.SourceGenerator/Generator.cs

```diff
@@ -1,258 +1,342 @@
+                        /// </summary>
+                        /// <param name="activatableClassId">The class identifier that is associated with an activatable runtime class.</param>
+                        /// <param name="factory">A pointer to the activation factory that corresponds with the class specified by <paramref name="activatableClassId"/>.</param>
+                        /// <returns>The <c>HRESULT</c> for the operation.</returns>
+                        /// <seealso href="https://learn.microsoft.com/en-us/previous-versions/br205771(v=vs.85)"/>
+                        [UnmanagedCallersOnly(EntryPoint = nameof(DllGetActivationFactory), CallConvs = new[] { typeof(CallConvStdcall) })]
+                        public static int DllGetActivationFactory(void* activatableClassId, void** factory)
+                        {
+                            const int E_INVALIDARG = unchecked((int)0x80070057);
+                            const int E_NOINTERFACE = unchecked((int)0x80004002);
+                            const int S_OK = 0;
+
+                            if (activatableClassId is null || factory is null)
+                            {
+                                return E_INVALIDARG;
+                            }
+
+                            try
+                            {
+                                IntPtr obj = GetActivationFactory(MarshalString.FromAbi(activatableClassId));
+
+                                if ((void*)obj is null)
+                                {
+                                    *factory = null;
+
+                                    return E_NOINTERFACE;
+                                }
+
+                                *factory = (void*)obj;
+
+                                return S_OK;
+                            }
+                            catch (Exception e)
+                            {
+                                ExceptionHelpers.SetErrorInfo(ex);
+
+                                return ExceptionHelpers.GetHRForException(ex);
+                            }
+                        }
+
+                        /// <summary>
+                        /// Determines whether the DLL that implements this function is in use. If not, the caller can unload the DLL from memory.
+                        /// </summary>
+                        /// <returns>This method always returns <c>S_FALSE</c>.</returns>
+                        /// <seealso href="https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-dllcanunloadnow"/>
+                        [UnmanagedCallersOnly(EntryPoint = nameof(DllCanUnloadNow), CallConvs = new[] { typeof(CallConvStdcall) })]
+                        public static int DllCanUnloadNow()
+                        {
+                            const int S_FALSE = 1;
+
+                            return S_FALSE;

```

Comment:
This will prevent the dll from unloading even if it's not used anymore, you should return true if the total ref count of implemented classes is zero, I'm not sure whether CsWinRT already has a ref tracker for this (similar to Module Lock in CppWinRT) or not tho.

Reply:
This matches the behavior of the native host. Also, note that NativeAOT doesn't support unloading anyway.